### ISSUE TITLE
Methods within preprocessor directives are not documented (C++)

### ISSUE THREAD

**Issue by alve89** [doxygen/doxygen#8694]
**Describe the bug**
Method which is placed within preprocessor directives is not documented.

```
class dogboxHandler : public RequestHandler {
  /**
   * Method to define which HTTP requests (= which addresses / files / targets) can be handled
   * @param  method   HTTP-method of the request (GET/POST)
   * @param  uri      Request as string
   * @return          Ability to handle the desired request
   *                   `True` if URI can be handled
   *                   `False` if URI can't be handled
   */
  bool canHandle(HTTPMethod method, String uri);
  


  /**
   * Handles the HTTP request if the URI was specified in canHandle()
   * @param  server         Server object, either of type `ESP8266Webserver` or `WebServer`, depending on the used architecture (ESP8266 or ESP32)
   * @param  requestMethod  Method of the HTTP request
   * @param  requestUri     Target of the request
   * @return                `True` if everything was processed properly, `false` if not
   */
  #if defined ESP8266
    bool handle(ESP8266WebServer& server, HTTPMethod requestMethod, String requestUri);
  #elif defined ESP32
    bool handle(WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
};
```

**Expected behavior**
I thought that this method is documented as well.

**Screenshots**
![image](https://user-images.githubusercontent.com/11487054/127275698-abe48aa9-a2e1-4c3f-8fb4-90a6b352bc0f.png)

**To Reproduce**
[Example.zip](https://github.com/doxygen/doxygen/files/6890859/Example.zip)

**Version**
1.9.1

**Additional context**
Even if the documentation block is placed directly before the method declaration it isn't documented

```
class dogboxHandler : public RequestHandler {
  /**
   * Method to define which HTTP requests (= which addresses / files / targets) can be handled
   * @param  method   HTTP-method of the request (GET/POST)
   * @param  uri      Request as string
   * @return          Ability to handle the desired request
   *                  `True` if URI can be handled
   *                  `False` if URI can't be handled
   */
  bool canHandle(HTTPMethod method, String uri);




  #if defined ESP8266
  /**
   * Handles the HTTP request if the URI was specified in canHandle()
   * @param  server         Server object, either of type `ESP8266Webserver` or `WebServer`, depending on the used architecture (ESP8266 or ESP32)
   * @param  requestMethod  Method of the HTTP request
   * @param  requestUri     Target of the request
   * @return                `True` if everything was processed properly, `false` if not
   */
    bool handle(ESP8266WebServer& server, HTTPMethod requestMethod, String requestUri);
  #elif defined ESP32
    bool handle(WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
};
```

**Comment by albert-github** [comment]
In your `Doxyfile` there is no indication that you set any `PREDEFINED` value. So the result is as expected, both `handle` functions are not present.

When you set:
```
MACRO_EXPANSION=YES
PREDEFINED = ESP8266
```
or
```
MACRO_EXPANSION=YES
PREDEFINED = ESP32
```
you will get the first respectively the second `handle` function. 


So I think the problem is usage.

**Comment by alve89** [comment]
Thank you for your quick reply! Your solution works as expected, but I can't get both `handle` functions (always only the first **or** the second)...

I tried this:
```
MACRO_EXPANSION        = YES
ENABLE_PREPROCESSING   = YES
```
with both
```
PREDEFINED             = ESP32 ESP8266
```
and
```
PREDEFINED             = ESP32
PREDEFINED             = ESP8266
```

How can I specify multiple tags?

**Comment by albert-github** [comment]
The definition with `PREDEFINED             = ESP32 ESP8266` is OK an alternative would be:
```
PREDEFINED             = ESP32
PREDEFINED             += ESP8266
```

The problem here is of course that you have here a `#if ... #elif ...` structure so always one of the 2 is selected (like the compiler would do). By means splitting the `#if ... #elif ...` into 2 independent `#if` constructs you can overcome this., but you have to write the documentation for both parts, also for this there is a way around:
```
  /**
    [handle]
     Handles the HTTP request if the URI was specified in canHandle()
     @param  server         Server object, either of type `ESP8266Webserver` or `WebServer`, depending on the used architecture (ESP8266 or ESP32)
     @param  requestMethod  Method of the HTTP request
     @param  requestUri     Target of the request
     @return                `True` if everything was processed properly, `false` if not
    [handle]
   */
  #if defined ESP8266
    bool handle(ESP8266WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
  /** \snippetdoc this handle */
  #if defined ESP32
    bool handle(WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
```
Note:
- the `*` in the description have been removed as otherwise the snippet will see them as list starters
- due to the double `#if` you run in the code the risk that both will be selected, but with some additional `#if` logic you could overcome this.

**Comment by alve89** [comment]
Thank you so much for this very detailled reply! It's working perfectly now!

Except of the displayed `[handle]` (I copied your code from above), but that's not problematic for me, I just use a specific documentation for each method.

![image](https://user-images.githubusercontent.com/11487054/127312527-3433e9e4-1a8d-445a-ac84-218d005e3679.png)

**Comment by albert-github** [comment]
Even though you have separate documentation for each function.

I see the problem here, workaround:
```
class dogboxHandler : public RequestHandler {
  /**
   * Method to define which HTTP requests (= which addresses / files / targets) can be handled
   * @param  method   HTTP-method of the request (GET/POST)
   * @param  uri      Request as string
   * @return          Ability to handle the desired request
   *                  `True` if URI can be handled
   *                  `False` if URI can't be handled
   */
  bool canHandle(HTTPMethod method, String uri);




  /** \snippetdoc this handle */
  #if defined ESP8266
    bool handle(ESP8266WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
  /** \snippetdoc this handle */
  #if defined ESP32
    bool handle(WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
};
  /**
    [handle]
     Handles the HTTP request if the URI was specified in canHandle()
     @param  server         Server object, either of type `ESP8266Webserver` or `WebServer`, depending on the used architecture (ESP8266 or ESP32)
     @param  requestMethod  Method of the HTTP request
     @param  requestUri     Target of the request
     @return                `True` if everything was processed properly, `false` if not
    [handle]
   */
```
so just place the snippet in a separate, not directly used, comment block.

**Comment by albert-github** [comment]
Originally the snippet was thought for code snippets and not for documentation parts, resulting in this a bit unwanted behavior.

@doxygen how about commands `\snippetstart (block_id)` and `\snippetend (block_id)`?

**Comment by alve89** [comment]
Hm, this solution is a problem for my "inner Monk" - every documentation blocks are before the method declaration, here it would be after the declaration...

**Comment by albert-github** [comment]
When using regular doxygen comments this is true, though as a workaround here also the version with standard comments `/* ... */` works, i.e.
```
class dogboxHandler : public RequestHandler {
  /**
   * Method to define which HTTP requests (= which addresses / files / targets) can be handled
   * @param  method   HTTP-method of the request (GET/POST)
   * @param  uri      Request as string
   * @return          Ability to handle the desired request
   *                  `True` if URI can be handled
   *                  `False` if URI can't be handled
   */
  bool canHandle(HTTPMethod method, String uri);


  /*
    [handle]
     Handles the HTTP request if the URI was specified in canHandle()
     @param  server         Server object, either of type `ESP8266Webserver` or `WebServer`, depending on the used architecture (ESP8266 or ESP32)
     @param  requestMethod  Method of the HTTP request
     @param  requestUri     Target of the request
     @return                `True` if everything was processed properly, `false` if not
    [handle]
   */


  /** \snippetdoc this handle */
  #if defined ESP8266
    bool handle(ESP8266WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
  /** \snippetdoc this handle */
  #if defined ESP32
    bool handle(WebServer& server, HTTPMethod requestMethod, String requestUri);
  #endif
};
```

it is a bit against all feelings that doxygen uses also information from non-doxygen comment blocks, but this is due to the fact that doxygen does a very simple search for detecting the block markers.

**Comment by albert-github** [comment]
Apparently an intriguing issue for me.

Regarding the `[handle]` problem here there is even an easier solution than creating a new command we can replace `[handle]` by `\noop [handle]`. The  snippet detector will pick this up properly  and in the documentation we can use the standard doxygen `/**` comment again, so all like:
```
  /**
    \noop [handle]
     Handles the HTTP request if the URI was specified in canHandle()
     @param  server         Server object, either of type `ESP8266Webserver` or `WebServer`, depending on the used architecture (ESP8266 or ESP32)
     @param  requestMethod  Method of the HTTP request
     @param  requestUri     Target of the request
     @return                `True` if everything was processed properly, `false` if not
    \noop [handle]
   */
  #if defined ESP8266
```
(Note the `\noop` command s available since version 1.8.17)

Regarding the `#if ...#elif` problem here is besides the splitting of the condition also a more doxgen like solution.
With the aid of the `\comd` command it is possible to remove certain parts of the comments **and** code.
In this case one could write:
```
  /** \cond */
  #if defined ESP8266
  /** \endcond */
    bool handle(ESP8266WebServer& server, HTTPMethod requestMethod, String requestUri);
  /** \cond */
  #elif defined ESP32
  /** \endcond */
  /** \snippetdoc this handle */
    bool handle(WebServer& server, HTTPMethod requestMethod, String requestUri);
  /** \cond */
  #endif
  /** \endcond */
```
the `\cond` "works  as" and `#if`, in case an id is given (`\=cond <id>`) the condition is `true` in case the `id` is set in the `ENABLED_SECTIONS` otherwise teh condition is `false` the condition is is also `false` when no `id` is given.
As with the workaround of splitting the `#if` one has to be careful, and it cann't / won't always work properly (e.g. with `#define`).
